process/track/shardBlockTrack.go

```diff
@@ -0,0 +1,135 @@
+	err := checkTrackerNilParameters(arguments.ArgBaseTracker)
+	if err != nil {
+		return nil, err
+	}
+
+	if check.IfNil(arguments.PoolsHolder) {
+		return nil, process.ErrNilPoolsHolder
+	}
+	if check.IfNil(arguments.PoolsHolder.Headers()) {
+		return nil, process.ErrNilHeadersDataPool
+	}
+	if check.IfNil(arguments.PoolsHolder.MetaBlocks()) {
+		return nil, process.ErrNilMetaBlocksPool
+	}
+	if check.IfNil(arguments.PoolsHolder.HeadersNonces()) {
+		return nil, process.ErrNilHeadersNoncesDataPool
+	}
+
+	crossNotarizer, err := NewBlockNotarizer(arguments.Hasher, arguments.Marshalizer)
+	if err != nil {
+		return nil, err
+	}
+
+	selfNotarizer, err := NewBlockNotarizer(arguments.Hasher, arguments.Marshalizer)
+	if err != nil {
+		return nil, err
+	}
+
+	crossNotarizedHeadersNotifier, err := NewBlockNotifier()
+	if err != nil {
+		return nil, err
+	}
+
+	selfNotarizedHeadersNotifier, err := NewBlockNotifier()
+	if err != nil {
+		return nil, err
+	}
+
+	bbt := &baseBlockTrack{
+		hasher:                        arguments.Hasher,
+		headerValidator:               arguments.HeaderValidator,
+		marshalizer:                   arguments.Marshalizer,
+		rounder:                       arguments.Rounder,
+		shardCoordinator:              arguments.ShardCoordinator,
+		shardHeadersPool:              arguments.PoolsHolder.Headers(),
+		metaBlocksPool:                arguments.PoolsHolder.MetaBlocks(),
+		headersNoncesPool:             arguments.PoolsHolder.HeadersNonces(),
+		store:                         arguments.Store,
+		crossNotarizer:                crossNotarizer,
+		selfNotarizer:                 selfNotarizer,
+		crossNotarizedHeadersNotifier: crossNotarizedHeadersNotifier,
+		selfNotarizedHeadersNotifier:  selfNotarizedHeadersNotifier,
+	}
+
+	err = bbt.initNotarizedHeaders(arguments.StartHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	sbt := shardBlockTrack{
+		baseBlockTrack: bbt,
+	}
+
+	blockProcessor, err := NewBlockProcessor(
+		arguments.HeaderValidator,
+		arguments.ShardCoordinator,
+		&sbt,
+		crossNotarizer,
+		crossNotarizedHeadersNotifier,
+		selfNotarizedHeadersNotifier,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	sbt.blockProcessor = blockProcessor
+
+	sbt.headers = make(map[uint32]map[uint64][]*headerInfo)
+	sbt.shardHeadersPool.RegisterHandler(sbt.receivedShardHeader)
+	sbt.metaBlocksPool.RegisterHandler(sbt.receivedMetaBlock)
+
+	return &sbt, nil
+}
+
+func (sbt *shardBlockTrack) getSelfHeaders(headerHandler data.HeaderHandler) []*headerInfo {
+	selfHeadersInfo := make([]*headerInfo, 0)
+
+	metaBlock, ok := headerHandler.(*block.MetaBlock)
+	if !ok {
+		log.Debug("getSelfHeaders", process.ErrWrongTypeAssertion)
+		return selfHeadersInfo
+	}
+
+	for _, shardInfo := range metaBlock.ShardInfo {
+		if shardInfo.ShardID != sbt.shardCoordinator.SelfId() {
+			continue
+		}
+
+		header, err := process.GetShardHeader(shardInfo.HeaderHash, sbt.shardHeadersPool, sbt.marshalizer, sbt.store)
```

Comment:
there is no way for this header to be in the poo. It is only in the store.

Reply:
If one node constructed on the fork in shard and now it received a meta block which contains the wright shard headers which have been notarized in meta, these shard headers would be only in pool as they were not been removed because node didn't use them as it constructed on the wrong chain.

Reply:
These shard headers contained in the received meta block are used to notify all the registered handlers (for now only fork detector) that there are new self headers notarized in meta, so the node can now check sooner if it constructed on the notarized chain or not (not need anymore to wait until it will include these meta headers in its processed shard headers to determine this)